### ISSUE TITLE
ZeroShotClassifier takes a loooong time

### ISSUE THREAD

**Issue by noahberhe** [MaartenGr/BERTopic#1159]
Any ideas how I could improve it, I'm only running on a couple thousand short documents

**Comment by doubianimehdi** [comment]
hum ... not you are in the right place ... bertopic is not a classifier ...

**Comment by MaartenGr** [comment]
@doubianimehdi That's not entirely true. As of the last 0.14 release, BERTopic supports a number of representation models, one of which is a [ZeroShotClassification](https://maartengr.github.io/BERTopic/getting_started/representation/representation.html#zero-shot-classification) module. Moreover, you can switch out the clustering algorithm with a classifier to do [supervised topic modeling](https://maartengr.github.io/BERTopic/getting_started/supervised/supervised.html). 

@noahberhe Could you share the code that you have been using? The `ZeroShotClassification` function can be fast or slow depending on the model that you use, including your own hardware.

**Comment by noahberhe** [comment]
Here you go... :)
And I'm using a CPU: 2 GHz Quad-Core Intel Core i5,
I also tried it with `valhalla/distilbart-mnli-12-1` and still took a while.

```
n_docs = 10000
comments_bert_train = df.admin_comment.sample(n=n_docs, random_state=1).values

# Pre-train embedding model
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(comments_bert_train, show_progress_bar=True)

# Representation model
candidate_topics = df_labels.labels.values
representation_model = ZeroShotClassification(candidate_topics, model="facebook/bart-large-mnli")

# Reduce frequent words
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)

topic_model = BERTopic(top_n_words=3,
                       representation_model=representation_model,
                       ctfidf_model=ctfidf_model,
                       embedding_model=sentence_model,
                       n_gram_range=(1, 2),
                       calculate_probabilities=False,
                       verbose=True)

topics, probs = topic_model.fit_transform(comments_bert_train, embeddings)
```

**Comment by MaartenGr** [comment]
You are trying to use a large language model on the CPU, which is typically not recommended. These generally require a good GPU to run, so I think that is why it is so slow for you.

**Comment by noahberhe** [comment]
Thanks, I might give it a go once I've got an AWS instance up and running.
But, so far I've got decent results with GuidedBERTopic, which is giving me a relatively nice bigram to describe each topic

```
n_docs = 10000
comments_bert_train = df.admin_comment.sample(n=n_docs, random_state=1).values

# Pre-train embedding model --- to improve speed
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(comments_bert_train, show_progress_bar=True)

representation_model = KeyBERTInspired(top_n_words=1)

ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)

seed_topic_list = df_labels.labels.values
seed_topic_list = [x.split() for x in seed_topic_list]

topic_model = BERTopic(ctfidf_model=ctfidf_model,
                       embedding_model=sentence_model,
                       representation_model=representation_model,
                       n_gram_range=(2, 2),
                       seed_topic_list=seed_topic_list,
                       calculate_probabilities=True,
                       verbose=True)

topics, probs = topic_model.fit_transform(comments_bert_train, embeddings)
```

**Comment by doubianimehdi** [comment]
oh my bad , I'm using setfit which has the same feature ... I didn't know that BERTopic had it ! sorry for the confusion !

**Comment by daniau23** [comment]
My question is how to extract the output just like when I am using the zeroshot classifier pipeline from hugging face
```
from bertopic.representation import ZeroShotClassification
from bertopic.vectorizers import ClassTfidfTransformer
from bertopic.dimensionality import BaseDimensionalityReduction

# sentence model and embeddings
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(docs, show_progress_bar=True)


# Skip over dimensionality reduction, replace cluster model with classifier,
# and reduce frequent words while we are at it.
empty_dimensionality_model = BaseDimensionalityReduction()
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)

# candidate topics and representation model
candidate_topics = ["usa and politics","ai and healthcare","sports","disaster", "nigeria and politics"]
representation_model = ZeroShotClassification(candidate_topics, model="facebook/bart-large-mnli")

# Create a fully supervised BERTopic instance
zero_shot_model= BERTopic(n_gram_range=(1,2),
                    embedding_model=sentence_model,
                    ctfidf_model=ctfidf_model,
                    umap_model=empty_dimensionality_model,
                    representation_model=representation_model,
                    verbose=True,
                    calculate_probabilities=False,
                                )
topics, probs = zero_shot_model.fit_transform(docs, embeddings)
```

**Comment by MaartenGr** [comment]
@daniau23 That seems like a reasonable pipeline. Is there something wrong with it?

**Comment by daniau23** [comment]
@MaartenGr I followed the example as presented by @noahberhe  and also followed the [zero shot link](https://maartengr.github.io/BERTopic/api/representation/zeroshot.html#bertopic.representation._zeroshot.ZeroShotClassification) 
But I'm failing to understand how to extract required output based off my candidate labels just as you would do if using the hugging face pipeline as shown in this [hugging face link](https://huggingface.co/facebook/bart-large-mnli)

I did try to use ```zero_shot_model.get_topic_info()```, but it's not consistent with my candidate labels. Is there something that I am missing?

**Comment by MaartenGr** [comment]
@daniau23 I am not sure if I understand you correctly. The zero shot approach in BERTopic essentially checks whether certain automatically generated topic labels, through c-TF-IDF, could be better represented with one candidate labels. If they could not, or if they do not match well, the original c-TF-IDF labels are used.

**Comment by daniau23** [comment]
I understand your point, thank for your explanation.
